src/AS_02_IAB.cpp

```diff
@@ -0,0 +1,920 @@
+    {
+        DefaultLogSink().Error("IABEssecenDescriptor.EditRate is not a supported value: %d/%d\n",
+            ADesc.FrameRate.Numerator, ADesc.FrameRate.Denominator);
+        return RESULT_RAW_FORMAT;
+    }
+
+    assert(m_Dict);
+
+    memcpy(m_EssenceUL, m_Dict->ul(MDD_IMF_IABEssenceClipWrappedElement), SMPTE_UL_LENGTH); 
+    m_EssenceUL[SMPTE_UL_LENGTH-1] = 1; // first (and only) essence container
+    Result_t result = m_State.Goto_READY();
+
+    if ( ASDCP_SUCCESS(result))
+    {
+        ui32_t TCFrameRate = (ui32_t)ceil(ADesc.FrameRate.Quotient());
+
+        // timecode-in-imf best practices requires that we do not use timecodes in IAB track files
+        SetUseTimecodes(0);
+
+        result = WriteAS02Header(
+            packageLabel, 
+            UL(m_Dict->ul(MDD_IMF_IABEssenceClipWrappedContainer)), //  Wrapping
+            trackName, 
+            UL(m_EssenceUL), // Essence UL
+            UL(m_Dict->ul(MDD_SoundDataDef)), // DataDefinition
+            ADesc.FrameRate, 
+            TCFrameRate);
+    }
+    if (KM_SUCCESS(result))
+    {
+        // even though the header was already written out, it will be 
+        // written out again at the end when footer is written
+        Batch<UL> specs;
+        specs.push_back(UL(m_Dict->ul(MDD_IMF_IABTrackFileLevel0)));
+        m_HeaderPart.m_Preface->ConformsToSpecifications.set(specs);
+    }
+
+    m_IndexWriter.SetEditRate(ADesc.FrameRate);
+
+    return result;
+}
+
+// Writes a frame of essence to the MXF file. If the optional AESEncContext
+// argument is present, the essence is encrypted prior to writing.
+// Fails if the file is not open, is finalized, or an operating system
+// error occurs.
+//
+Result_t AS_02::IAB::MXFWriter::h__Writer::WriteFrame(const ASDCP::FrameBuffer& frame_buf,
+    AESEncContext* Ctx, HMACContext* HMAC)
+{
+    if (frame_buf.Size() == 0)
+    {
+        DefaultLogSink().Error("The frame buffer size is zero.\n");
+        return RESULT_PARAM;
+    }
+
+    Result_t result = RESULT_OK;
+
+    if (m_State.Test_READY())
+    {
+        result = m_State.Goto_RUNNING(); // first time through
+    }
+
+    
+
+    if (KM_SUCCESS(result) && ! HasOpenClip())
+    {
+        ui64_t position = m_File.Tell();
+        result = StartClip(m_EssenceUL, Ctx, HMAC);
+        m_StartPayloadOffset = ui32_t(m_File.Tell() - position);
+    }
+
+    if (KM_SUCCESS(result))
+    {
+        result = WriteClipBlock(frame_buf);
+    }
+
+    if ( ASDCP_SUCCESS(result) )
+    {
+        IndexTableSegment::IndexEntry Entry;
+        Entry.StreamOffset = m_TotalBytesWritten + m_StartPayloadOffset;
+
+        m_IndexWriter.PushIndexEntry(Entry);
+
+        m_FramesWritten ++;
+        m_TotalBytesWritten += frame_buf.Size();
+    }
+
+    return result;
+}
+
+Result_t AS_02::IAB::MXFWriter::h__Writer::WriteMetadata(const std::string &trackLabel, const std::string &mimeType, const std::string &dataDescription, const ASDCP::FrameBuffer& metadata_buf)
```

Comment:
Note: this function reimplements SMPTE RP-2057, which was implemented as part of `AddDmsTrackGenericPartUtf8Text` meanwhile.
Ultimately, they could maybe be merged together if you believe that's better, but they have slightly different behavior, so for now it might be better to keep them as is, maybe adding a note/TODO about this here.